KEP-3352: Aggregated Discovery KEP to Alpha (#3364)
kubernetes/enhancements: keps/sig-api-machinery/3352-aggregated-discovery/README.md

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

> ## Summary

Reviewer: forgot this section?

possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

> All clients and users of Kubernetes APIs usually first need to “discover” what the available APIs are and how they can be used. These APIs are described through a mechanism called “Discovery” which is typically queried to then build the requests to correct APIs. Unfortunately, the “Discovery” API is made of lots of small objects that need to be queried individually, causing possibly a lot of delay due to the latency of each individual request (up to 80 requests, with most objects being less than 1,024 bytes). The more complex the APIs provided by the Kubernetes cluster, the more requests need to be performed.

Reviewer: minor: mind putting line breaks at every sentence so it's easy to know what I'm commenting on?

We will add a new controller responsible for aggregating the discovery documents when a resource on the cluster changes. There will be no conflicts when aggregating since each discovery document is self-contained.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

> We will expose a endpoint `/discovery` that will support JSON, and

Reviewer: how will we version this API so that we don't burn the name when we (probably) need a v3?

Author: How about something like `/discovery/v1` similar to how we version the OpenAPI?

Reviewer: > How about something like /discovery/v1 similar to how we version the OpenAPI?

That would work for me.

> We will expose a endpoint `/discovery` that will support JSON, and

Reviewer: will we also support compression?

Author: We currently do not support gzip for the `/api` and `/apis` endpoint. I guess it wouldn't hurt to add support for gzip here

protobuf. The endpoint will serve the aggregated discovery document
for all types that a Kubernetes cluster supports.

### API

> The contents of this endpoint will be an `APIGroupList`, which is the

Reviewer: are you open to removing fields not related to discovery or that we have found limited use for?  `serverAddressByClientCIDRs` comes to mind as "wish we hadn't tried that"

Author: I'm assuming we can't deprecate the GA field and we would have to create a new struct without that field? I don't really have a strong preference, can add it to the KEP if we'd like to create new discovery structs.

Reviewer: I think we should make our API serialization the type we want. I can understand if internally you have a RESTMapper convert to something else to make code easier to migrate.

Reviewer 2: I agree with David, we can make a type that's a subset of the existing type and only includes necessary fields. We're not removing the existing pages, people can go there if they really need to.

Reviewer 3: Did this make it into the merged KEP?  I don’t see the language in the merged version I expected (a new Kind, with fewer fields, and discussion of naming).

We have a mechanism to request a specific variant of a type (content negotiation), so I would prefer we use a kind that requires minimal changes and clarify what happens if you need those fields, not bifurcate the types.

Author: We're creating a new API in a new endpoint (`discovery/v1`) with a different API format compared to `/apis` and `apis/<group>/<version>`. The types are not round-trippable between the two APIs (eg: Subresource return types are changed, etc) and serve different endpoints which is why we ended up deciding to create new types.

Feel free to comment directly in the PR:  https://github.com/kubernetes/kubernetes/pull/111978

restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->
Yes, the feature may be disabled by reverting the feature flag.

###### What happens if we reenable the feature if it was previously rolled back?

The feature does not depend on state, and can be disabled/enabled at will.

###### Are there any tests for feature enablement/disablement?

<!--
The e2e framework does not currently support enabling or disabling feature
gates. However, unit tests in each component dealing with managing data, created
with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.

Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

> n/a

Reviewer: I'd expect a test that makes sure that the kubectl discovery/RESTMapper logic functions properly when the endpoint is both disabled and enabled.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

> ###### What specific metrics should inform a rollback?

Reviewer: This seems worthwhile to address.  Failures of a metric in the aggregation controller perhaps?  Failure codes from the /discovery endpoint itself?

and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

- [ ] Events
  - Event Reason:
- [ ] API .status
  - Condition name:
  - Other field:
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

> - [ ] Metrics

Reviewer: While not strictly required for alpha, I think filling this section out may give confidence to @tkashem and @soltysh if they wanted to try enabling it.  Without it, I think they would not be able to try the alpha at any kind of scale.

Reviewer 2: Could also put in the readyz label here maybe.

Metrics could be like, how many times the etag has changed, and how long it took to do the aggregation.

  - [Optional] Aggregation method:
  - Components exposing the metric:
- [ ] Other (treat as last resort)
  - Details:

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

<!--
Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
implementation difficulties, etc.).
-->

### Dependencies

<!--
This section must be completed when targeting beta to a release.
-->

###### Does this feature depend on any specific services running in the cluster?

<!--
Think about both cluster-level services (e.g. metrics-server) as well
as node-level agents (e.g. specific version of CRI). Focus on external or
optional services that are needed. For example, if this feature depends on
a cloud provider API, or upon an external software-defined storage or network
control plane.

For each of these, fill in the following—thinking about running existing user workloads
and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

> ### Scalability

Reviewer 4: Do you have an estimate as to what kind of performance increase we might see with this change? Is that something that you'd measure to determine how successful the new feature was (i.e. whether it was worth taking to beta etc)?

Per https://github.com/crossplane/crossplane/blob/60fc7d/design/one-pager-crd-scaling.md#client-side-throttling we currently see discovery taking ~20 seconds on EKS with ~360 fairly large API groups.

Reviewer 2: (Turn off client side rate limiting and/or check your APF settings to solve in the mean time.) This should eliminate one source of that kind of problem, I can't promise you don't have others.

Reviewer 4: @lavalamp I appreciate the input but it's not feasible for us as the maintainers of Crossplane to expect our community to go tune this. I believe even turning off client side rate limiting for discovery in kubectl (among other clients) currently requires forking - I don't think it's exposed as a flag.

I intuitively expect this KEP to result in much more performant discovery - what I'm curious about is whether you have any idea how much more performant it will be.

Author: @negz There are a variety of factors that affect the performance of discovery (latency between client-server, throughput, rate limiting, etc) so it's hard for us to give an exact answer. 

With this KEP, the total number of requests will be reduced from `kubectl api-versions | wc -l` to a single request. The total size of the content downloaded will be the exact same as the aggregate size of the current content downloaded via separate GVs (perhaps the aggregated document will be slightly smaller after introducing compression). Burst limits will no longer be a problem as only one request will be sent to fetch all the group-versions.

      PRs/issues, release notes

<!-- **Note:** This checklist is iterative and should be reviewed and
updated every time this enhancement is being considered for a
milestone. -->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

<!-- This section is incredibly important for producing high-quality,
user-focused documentation such as release notes or a development
roadmap. It should be possible to collect this information before
implementation begins, in order to avoid requiring implementors to
split their attention between writing release notes and implementing
the feature itself. KEP editors and SIG Docs should help to ensure
that the tone and content of the `Summary` section is useful for a
wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the
[documentation style guide]. In particular, wrap lines to a reasonable
length, to make it easier for reviewers to cite specific portions, and
to minimize diff churn on updates.
-->

> The operations that a Kubernetes API server support is dictated

Reviewer 2: ```suggestion
The operations that a Kubernetes API server supports are reported
```

storm of requests for clusters and is a source of latency and
throttling. This KEP proposes centralizing the "discovery" mechanism
into one document so that clients only need to make one request to the
API server to retrieve all the operations available.

## Motivation

<!-- This section is for explicitly listing the motivation, goals, and
non-goals of this KEP. Describe why the change is important and the
benefits to users. The motivation section can optionally provide links
to [experience reports] to demonstrate the interest in a KEP within
the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports -->

All clients and users of Kubernetes APIs usually first need to
“discover” what the available APIs are and how they can be used. These
APIs are described through a mechanism called “Discovery” which is
typically queried to then build the requests to correct APIs.
Unfortunately, the “Discovery” API is made of lots of small objects
that need to be queried individually, causing possibly a lot of delay
due to the latency of each individual request (up to 80 requests, with

> most objects being less than 1,024 bytes). The more complex the APIs

Reviewer 2: ```suggestion
most objects being less than 1,024 bytes). The more numerous the APIs
```

provided by the Kubernetes cluster, the more requests need to be
performed.

The most well known Kubernetes client that uses the discovery
mechanism is `kubectl`, and more specifically the
`CachedDiscoveryClient` in `client-go`. To mitigate some of this

> latency, kubectl has implemented a (10 minute timer) 6 hours(!) timer

Reviewer 2: 10 minutes or 6 hours?

Reviewer 4: It was recently bumped from 10 mins to 6 hours in https://github.com/kubernetes/kubernetes/pull/107141.

Author: Copying strikethrough text from google docs to markdown clearly shouldn't be attempted. Should say 6 hours.

during which the discovery API is not refreshed. The drawback of this
approach is that the freshness of the cache is doubtful and the entire
discovery API needs to be refreshed after 6 hours, even if it hasn’t
expired.

This not only impacts kubectl, but all clients of kubernetes. We can
do better.

### Goals

- Fix the discovery storm issue currently present in kubectl
- Aggregate the discovery documents for all Kubernetes types

<!-- List the specific goals of the KEP. What is it trying to achieve?
How will we know that this has succeeded? -->

### Non-Goals

<!-- What is out of scope for this KEP? Listing non-goals helps to
focus discussion and make progress. -->

- Since the current discovery separated by group-version is already

>   GA, removal of the endpoint will be difficult. This KEP will solely

Reviewer 2: ```suggestion
  GA, removal of the endpoint will not be attempted. This KEP will solely
```

  focus on introducing the new aggregated endpoint and will not cover

>   deprecation.

Reviewer 2: This slightly implies deprecation is the same as removal, which it isn't.

of your change are understandable. This may include API specs (though
not always required) or even code snippets. If there's any ambiguity
about HOW your proposal will be implemented, this is the place to
discuss them. -->

We will expose a endpoint `/discovery` that will support JSON, and
protobuf and gzip compression. The endpoint will serve the aggregated
discovery document for all types that a Kubernetes cluster supports.

### API

The contents of this endpoint will be an `APIGroupList`, which is the
same type that is returned in the discovery document at `/apis`. The
`APIGroupList` contains a list of `APIGroup` types which includes a
list of
[`GroupVersionForDiscovery`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1071)
types. We will modify this `GroupVersionForDiscovery` type to include
a list of
[`APIResource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1080).
This list of `APIResource` is what is currently published at the
`/apis/<group>/version` endpoint and is what will be aggregated into

> the new endpoint that we publish.

Reviewer 2: I think we can omit storageVersionHash and just let the storage version migrator keep looking at the old place?

Reviewer 3: What's the guideline that says whether we can omit something?  I'm a little surprised at the mindset of having a new API and saying "no no old client, you have to keep using it the old way".  Why is this new API not the best way for *any* client to get discovery vs the manual crawl?

Reviewer 2: I don't think SVM wants to see cached data.

Reviewer 2: Another reason the old version will never go away is that the aggregator will be reading the old version off of aggregated apiservers to produce the new cached version.

Would it be less confusing if we picked a new name rather than a new version?

Reviewer 3: Is there no way to hit `/discovery/v1` and cache bust?  Or are we so committed to a controller design that that's the key outcome?

Now that I look, I don't see anything in the KEP that bounds the latency of the doc - could we specify that in the design (we accept X latency in the absence of errors)?

Your point about SVM not wanting to see cached results... means that someone who is *mostly* ok with current results but occasionally needs to force that has no tool for it?  If SVM has a previous version of the results, and the aggregated APIs are down, as long as the new entry point can bound the cached endpoint on the etag, or omit it and get "the most recently bounded copy" (if it were seconds), is that really a problem for SVM?  Could we simply let the request for etag wait for the next controller interval to run (if short) such that it was always 'most recent / bounded'?

There seems to be a lot of expectations in the design of this that are focused on kubectl, but that's just one of our clients, and we *did* say "all clients" in the KEP...

Reviewer 3: After chatting - the analogy here is webhooks and how we fail closed if we can't refresh our config in a bounded interval.  Clients of this API can assume it is slightly delayed, but if something goes wrong in the controller they have no way of knowing how delayed, which (for the namespace example, is potentially catastrophic).  I don't think that's a requirement for alpha, but it's potentially a requirement for beta to ensure clients who may have ambiguous use cases don't a) accidentally use this api and regress, b) identify how we solve for it.

The endpoint will also publish an ETag calculated based on a hash of
the data for clients.

### Aggregation

> For the aggregation layer on the server, a new controller will be

Reviewer: This will likely need to have an input to readiness and you'll need to filter the discovery results by applicable group/version.

Reviewer 2: Yeah it's real bad if the first few clients get an empty doc because this controller hasn't run yet, but it's also bad to delay the cluster startup.

Reviewer 5: If api server signals it is ready even when discovery aggregation is not completed, this will not only affects clients but also causes flakiness of some e2e tests depends on discovery?

Reviewer: > If api server signals it is ready even when discovery aggregation is not completed, this will not only affects clients but also causes flakiness of some e2e tests depends on discovery?

More importantly, that causes bad behavior of the namespace lifecycle controller.

Reviewer 2: ah yeah that's another thing to consider, how does this affect that controller.
* will clients be able to know that e.g. a specific GV's backing apiserver hasn't been responsive for a while?
* if an apiserver restarts in that situation, how does it build the cache?

Author: Added that this will be an input to readiness.

For the controllers, afaik the behavior is based on the rest mapper. I've talked to @apelisse and I think we could keep the rest mapper behavior for controllers untouched and still have them use the old unaggregated endpoints. A single controller generally only cares about a very small subset of resources and should not need the entire discovery document. Since we will not be removing the old endpoint, we can optimize the existing controllers to only query for gvs that they need and have a new restmapper that uses the aggregated document.

  availability
- when the KEP was retired or superseded -->

## Drawbacks

With aggregation, the size of the aggregated discovery document could
be an issue in the future since clients will need to download the
entire document on any resource update. At the moment, even with 3000
CRDs (already very unlikely), the total size is still smaller than
1MB.

## Alternatives

<!-- What other approaches did you consider, and why did you rule them
out? These do not need to be as detailed as the proposal, but should
include enough information to express the idea and why it was not
acceptable. -->

An alternative that was considered is [Discovery Cache
Busting](https://docs.google.com/document/d/1AulBtUYjVcc4s809YSQq4bdRdDO3byY7ew9za4Ortj4).
Cache-busting allows clients to know if the files need to be
downloaded at all, and in most cases the download can be skipped
entirely. This typically works by including a hash of the resource
content in its name, while marking the objects as never-expiring using
cache control headers. Clients can then recognize if the names have
changed or stayed the same, and re-use files that have kept the same

> name without downloading them again.

Reviewer 2: Add TLDR on why we didn't choose this?

approach is that the freshness of the cache is doubtful and the entire
discovery API needs to be refreshed after 6 hours, even if it hasn’t
expired.

This not only impacts kubectl, but all clients of kubernetes. We can
do better.

### Goals

- Fix the discovery storm issue currently present in kubectl
- Aggregate the discovery documents for all Kubernetes types

<!-- List the specific goals of the KEP. What is it trying to achieve?
How will we know that this has succeeded? -->

### Non-Goals

<!-- What is out of scope for this KEP? Listing non-goals helps to
focus discussion and make progress. -->

- Since the current discovery separated by group-version is already
  GA, removal of the endpoint will be difficult. This KEP will solely
  focus on introducing the new aggregated endpoint and will not cover
  deprecation.

## Proposal

We are proposing a new endpoint `/discovery/v1` as an aggregated
endpoint for all discovery documents. Discovery documents can
currently be found under `apis/<group>/<version>` and `/api/v1` for
the legacy group version. This discovery endpoint will support
publishing an ETag so clients who already have the latest version of

> the aggregated discovery can avoid redownloading the document.

Reviewer 5: Does that mean that client will no longer cache documents as files?

Reviewer 5: or caching just one file with TTL

discuss them. -->

We will expose a endpoint `/discovery` that will support JSON, and
protobuf and gzip compression. The endpoint will serve the aggregated
discovery document for all types that a Kubernetes cluster supports.

### API

The contents of this endpoint will be an `APIGroupList`, which is the
same type that is returned in the discovery document at `/apis`. The
`APIGroupList` contains a list of `APIGroup` types which includes a
list of
[`GroupVersionForDiscovery`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1071)
types. We will modify this `GroupVersionForDiscovery` type to include
a list of
[`APIResource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1080).
This list of `APIResource` is what is currently published at the
`/apis/<group>/version` endpoint and is what will be aggregated into
the new endpoint that we publish.

The endpoint will also publish an ETag calculated based on a hash of
the data for clients.

This is approximately what the new API will look like (conflicting names will be renamed)

> ```go

Reviewer: looks good in concept.

// APIGroupList is a list of APIGroup, to allow clients to discover the API at
// /apis.
type APIGroupList struct {
	TypeMeta `json:",inline"`
	// groups is a list of APIGroup.
	Groups []APIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`
}

// APIGroup contains the name, the supported versions, and the preferred version
// of a group.
type APIGroup struct {
	TypeMeta `json:",inline"`
	// name is the name of the group.
	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
	// versions are the versions supported in this group.
	Versions []GroupVersionForDiscovery `json:"versions" protobuf:"bytes,2,rep,name=versions"`
	// preferredVersion is the version preferred by the API server, which
	// probably is the storage version.
	// +optional

> 	PreferredVersion GroupVersionForDiscovery `json:"preferredVersion,omitempty" protobuf:"bytes,3,opt,name=preferredVersion"`

Reviewer: Omit this and rely on the order in Versions?  A priority as opposed to a single version makes sense in light of sparse API group versions.

 	// ServerAddresssByClientCIDRs will be removed for the new Discovery API
	//ServerAddressByClientCIDRs []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs,omitempty" protobuf:"bytes,4,rep,name=serverAddressByClientCIDRs"`
}

// GroupVersion contains the "group/version" and "version" string of a version.
// It is made a struct to keep extensibility.
type GroupVersionForDiscovery struct {
	// groupVersion specifies the API group and version in the form "group/version"

> 	GroupVersion string `json:"groupVersion" protobuf:"bytes,1,opt,name=groupVersion"`

Reviewer: unnecessary?

Author: I think either a group or this groupVersion string would be useful here if we are just manipulating this one object without the parent APIGroup object.

Reviewer 6: I'm not a big fan of redundancy though

discovery document for all types that a Kubernetes cluster supports.

### API

The contents of this endpoint will be an `APIGroupList`, which is the
same type that is returned in the discovery document at `/apis`. The
`APIGroupList` contains a list of `APIGroup` types which includes a
list of
[`GroupVersionForDiscovery`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1071)
types. We will modify this `GroupVersionForDiscovery` type to include
a list of
[`APIResource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1080).
This list of `APIResource` is what is currently published at the
`/apis/<group>/version` endpoint and is what will be aggregated into
the new endpoint that we publish.

The endpoint will also publish an ETag calculated based on a hash of
the data for clients.

This is approximately what the new API will look like (conflicting names will be renamed)

```go
// APIGroupList is a list of APIGroup, to allow clients to discover the API at
// /apis.
type APIGroupList struct {
	TypeMeta `json:",inline"`
	// groups is a list of APIGroup.

> 	Groups []APIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`

Reviewer 2: is there a reason to have the group level? Why not just have all group versions?

Author: We need the separation so we can indicate the priority of group-versions within a group.

Reviewer 2: That field is commented out below, is that intentional?

Author: Yes, per David's comment, we'll remove the preferredVersion field and sort the Versions list based on priority. The preferredVersion should be at the top of the list.

Author: Oh, if you mean that the GroupVersion field is commented out, this was based on https://github.com/kubernetes/enhancements/pull/3364#discussion_r899326172

Reviewer 2: Well, you can still keep it sorted even if the list was flat, but we can work this out in the code.

All clients and users of Kubernetes APIs usually first need to
“discover” what the available APIs are and how they can be used. These
APIs are described through a mechanism called “Discovery” which is
typically queried to then build the requests to correct APIs.
Unfortunately, the “Discovery” API is made of lots of small objects
that need to be queried individually, causing possibly a lot of delay
due to the latency of each individual request (up to 80 requests, with
most objects being less than 1,024 bytes). The more numerous the APIs
provided by the Kubernetes cluster, the more requests need to be
performed.

The most well known Kubernetes client that uses the discovery
mechanism is `kubectl`, and more specifically the
`CachedDiscoveryClient` in `client-go`. To mitigate some of this
latency, kubectl has implemented a 6 hour timer during which the
discovery API is not refreshed. The drawback of this approach is that
the freshness of the cache is doubtful and the entire discovery API
needs to be refreshed after 6 hours, even if it hasn’t expired.

This not only impacts kubectl, but all clients of kubernetes. We can

> do better.

Reviewer 3: Some of the nuances we discussed in comments deserve to be in the KEP:

1. Which clients are we targeting primarily: clients that need to frequently request or maintain a discovery cache, like kubectl or web interfaces
2. We want all clients to benefit from this, but our alpha is targeted at solving these problems for those key consumers
3. We have some use cases that may require additional work to correctly support, we should identify those before entry to beta / reach consensus on their utility:
    * Namespace controller needs to ensure that discovery documents are refreshed after the namespace goes into the terminating phase (needs to guarantee that it sees the list of resources that could have been created within the namespace before the namespace is deleted).  The controller needs to be able to guarantee discovery doc list happens-after that, which is best solved by letting the namespace controller "wait" for the next refresh.  What we can't allow is for the namespace controller to not observe that resource
     * The use of server side caching may break *some* client side batch loops done with kubectl that need happens-before semantics, we should consider that usecase in kubectl and ensure we have a mitigation
4. Polling being sufficient over watch for most users
5. The API types in meta/v1 should be meta/v1alpha1 to start, not meta/v1

 	// ServerAddresssByClientCIDRs will be removed for the new Discovery API
	//ServerAddressByClientCIDRs []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs,omitempty" protobuf:"bytes,4,rep,name=serverAddressByClientCIDRs"`
}

// GroupVersion contains the "group/version" and "version" string of a version.
// It is made a struct to keep extensibility.
type GroupVersionForDiscovery struct {
	// groupVersion specifies the API group and version in the form "group/version"
	// This will be removed for the new discovery
	// GroupVersion string `json:"groupVersion" protobuf:"bytes,1,opt,name=groupVersion"`
	// version specifies the version in the form of "version". This is to save
	// the clients the trouble of splitting the GroupVersion.
	Version string `json:"version" protobuf:"bytes,2,opt,name=version"`
	// resources contains the name of the resources and if they are namespaced.
	APIResources []APIResource `json:"resources" protobuf:"bytes,2,rep,name=resources"`

 	// LastContacted is the last time that the apiserver has successfully reached the
 	// corresponding group version's discovery document. This will be nil if the group-version

>  	// has not been aggregated yet. To maintain consistency across scenarios with multiple

Reviewer 2: If it is nil, does that imply that APIResources above is an empty list?

Author: Yes, added a note

>  	// control planes, this time will be quantized down to the nearest fifteen minutes.

Reviewer 2: ```suggestion
 	// apiservers, this time will be quantized down to the nearest fifteen minutes.
```

(nit, sorry)

	// update, patch, delete, deletecollection, and proxy)
	Verbs Verbs `json:"verbs" protobuf:"bytes,4,opt,name=verbs"`
	// shortNames is a list of suggested short names of the resource.
	ShortNames []string `json:"shortNames,omitempty" protobuf:"bytes,5,rep,name=shortNames"`
	// categories is a list of the grouped resources this resource belongs to (e.g. 'all')
	Categories []string `json:"categories,omitempty" protobuf:"bytes,7,rep,name=categories"`

 	// StorageVersionHash will be removed in the new Discovery API
	// StorageVersionHash string `json:"storageVersionHash,omitempty" protobuf:"bytes,10,opt,name=storageVersionHash"`
}
```

### Aggregation

For the aggregation layer on the server, a new controller will be
created to aggregate discovery for built-in types, apiextensions types
(CRDs), and types from aggregated api servers.

A post start hook will be added and the kube-apiserver health check
should only pass if the discovery document is ready. Since aggregated
api servers may take longer to respond and we do not want to delay
cluster startup, the health check will only block on the local api
servers (built-ins and CRDs) to have their discovery ready. For api

> servers that have not been aggregated, their group-versions will be

Reviewer 3: Is this a typo?  "have not been aggregated"?  I expected to see a description of what client visible behavior we'd see for the discovery doc on initialization:

1. Built-ins - always visible?
2. CRDs - only visible once CRDs are loaded, so might be visible?
3. Aggregated APIs - could take materially longer to be visible?

I'd like to see some examples added to the KEP on the outcome for clients to clarify what they see during startup and how they know discovery is incomplete.  Will an incomplete discovery doc return a non 200 error message?  Will clients be expected to retry if the document is not complete?
